mlflow/gateway/providers/openai.py

```diff
@@ -0,0 +1,198 @@
+from typing import Dict, Any
+
+from fastapi.encoders import jsonable_encoder
+
+from .base import BaseProvider
+from ..schemas import chat, completions, embeddings
+
+
+class OpenAIProvider(BaseProvider):
+    NAME = "openai"
+    SUPPORTED_ROUTES = ("chat", "completions", "embeddings")
+
+    async def _request(self, path: str, payload: Dict[str, Any]):
+        import aiohttp
+
+        config = self.config.model.config
+        token = config["openai_api_key"]
+        headers = {"Authorization": f"Bearer {token}"}
+        if org := config.get("openai_organization"):
+            headers["OpenAI-Organization"] = org
+        async with aiohttp.ClientSession(headers=headers) as session:
+            url = "/".join([config["openai_api_base"].rstrip("/"), path.lstrip("/")])
+            async with session.post(url, json=payload) as response:
+                response.raise_for_status()
+                return await response.json()
+
+    @staticmethod
+    def _make_payload(payload: Dict[str, Any], mapping: Dict[str, str]):
+        payload = payload.copy()
+        for k1, k2 in mapping.items():
+            if v := payload.pop(k1, None):
+                payload[k2] = v
+        return {k: v for k, v in payload.items() if v is not None and v != []}
+
+    async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
+        payload = jsonable_encoder(payload)
+        if "n" in payload:
+            raise ValueError("Invalid parameter `n`. Use `candidate_count` instead.")
```

Comment:
What error type should we throw? `MlflowException.invalid_parameter`?

Reply:
Yep!